[samsung] Add climate IR samsung support

jorofi opened the pull request:
# What does this implement/fix?

New Feature: Samsung AC IR Control Component
I’ve added a custom component to enable seamless control of Samsung air conditioners using IR commands. With this component, you can:

Turn your Samsung AC on/off.
Adjust the temperature.
Change fan speeds and modes.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
https://github.com/esphome/esphome.io/pull/6862

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml

climate:
  - platform: samsung
    name: "Samsung AC"
    sensor: room_temperature
    receiver_id: rcvr

remote_receiver:
  id: rcvr
  pin:
    number: GPIO15
    inverted: true
    mode:
      input: true
      pullup: true
  tolerance: 55%

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DjordjeMandic commented:
Maybe component name / platform name should be `samsung` instead of `climate_ir_samsung`. For example like [`whirpool`](https://github.com/esphome/esphome/tree/dev/esphome/components/whirlpool)

DjordjeMandic reviewed (changes requested):
Component name / platform name should be `samsung` instead of `climate_ir_samsung`. Take a look at [`whirpool`](https://github.com/esphome/esphome/tree/dev/esphome/components/whirlpool) for example

jorofi commented:
> Component name / platform name should be `samsung` instead of `climate_ir_samsung`. Take a look at [`whirpool`](https://github.com/esphome/esphome/tree/dev/esphome/components/whirlpool) for example

I have renamed the component and I have also updated the documentation.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+  auto transmit = this->transmitter_->transmit();
+  auto *data = transmit.get_data();
+  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY);
+
```
```suggestion
data->reserve(<size>);
```
Take a look at nec_protocol.cpp for example at line 26-27 https://github.com/esphome/esphome/pull/8015

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,204 @@
+#define GETBITS8(data, offset, size) (((data) & (((uint8_t) UINT8_MAX >> (8 - (size))) << (offset))) >> (offset))
+
+static const char *const TAG = "samsung.climate";
+static const uint32_t SAMSUNG_IR_FREQUENCY = 38000;
```
```suggestion
static const uint32_t SAMSUNG_IR_FREQUENCY_HZ = 38000;
```

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,204 @@
+
+static const char *const TAG = "samsung.climate";
+static const uint32_t SAMSUNG_IR_FREQUENCY = 38000;
+static const int SAMSUNG_AIRCON1_HDR_MARK = 3000;
+static const int SAMSUNG_AIRCON1_HDR_SPACE = 9000;
+static const int SAMSUNG_AIRCON1_BIT_MARK = 500;
+static const int SAMSUNG_AIRCON1_ONE_SPACE = 1500;
+static const int SAMSUNG_AIRCON1_ZERO_SPACE = 500;
+static const int SAMSUNG_AIRCON1_MSG_SPACE = 2000;
```
```suggestion
// Timing constants in microseconds
static const uint32_t SAMSUNG_AIRCON1_HDR_MARK_US = 3000;
static const uint32_t SAMSUNG_AIRCON1_HDR_SPACE_US = 9000;
static const uint32_t SAMSUNG_AIRCON1_BIT_MARK_US = 500;
static const uint32_t SAMSUNG_AIRCON1_ONE_SPACE_US = 1500;
static const uint32_t SAMSUNG_AIRCON1_ZERO_SPACE_US = 500;
static const uint32_t SAMSUNG_AIRCON1_MSG_SPACE_US = 2000;
```

I think they're in positive microseconds and its fine to keep them as uint32_t. Is there specific reason not to do it? 

https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L27-L32

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,204 @@
+static const int SAMSUNG_AIRCON1_ZERO_SPACE = 500;
+static const int SAMSUNG_AIRCON1_MSG_SPACE = 2000;
+
+const uint16_t K_SAMSUNG_AC_EXTENDED_STATE_LENGTH = 21;
```
Maybe move this above `static const uint8_t K_RESET[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH]`

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+/// Set the temperature.
+/// @param[in] temp The temperature in degrees celsius.
+void SamsungClimate::set_temp_(const uint8_t temp) {
+  uint8_t new_temp = std::max(K_SAMSUNG_AC_MIN_TEMP, temp);
+  new_temp = std::min(K_SAMSUNG_AC_MAX_TEMP, new_temp);
+  protocol_.Temp = new_temp - K_SAMSUNG_AC_MIN_TEMP;
```
```suggestion
  protocol_.Temp = esphome::clamp<uint8_t>(temp, K_SAMSUNG_AC_MIN_TEMP, K_SAMSUNG_AC_MAX_TEMP);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+#include "samsung.h"
+
```
```suggestion
#include "samsung.h"
#include "esphome/core/helpers.h"

```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+    case climate::ClimateSwingMode::CLIMATE_SWING_BOTH:
+      protocol_.Swing = K_SAMSUNG_AC_SWING_BOTH;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_HORIZONTAL:
+      protocol_.Swing = K_SAMSUNG_AC_SWING_H;
+      ;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_VERTICAL:
+      protocol_.Swing = K_SAMSUNG_AC_SWING_V;
+      ;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_OFF:
+    default:
+      protocol_.Swing = K_SAMSUNG_AC_SWING_OFF;
+      ;
```
Any specific reason for extra `;`?

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY);
+
+  // Header
+  data->mark(SAMSUNG_AIRCON1_HDR_MARK);
+  data->space(SAMSUNG_AIRCON1_HDR_SPACE);
```
```suggestion
  data->item(SAMSUNG_AIRCON1_HDR_MARK_US, SAMSUNG_AIRCON1_HDR_SPACE_US);
```
https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L29-L32

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  for (int i = 0; i < 21; i++) {
+    if (i == 7 || i == 14) {
+      data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+      data->space(SAMSUNG_AIRCON1_MSG_SPACE);
```
```suggestion
      data->item(SAMSUNG_AIRCON1_BIT_MARK_US, SAMSUNG_AIRCON1_MSG_SPACE_US);
```
https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L29-L32

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+      data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+      data->space(SAMSUNG_AIRCON1_MSG_SPACE);
+
+      data->mark(SAMSUNG_AIRCON1_HDR_MARK);
+      data->space(SAMSUNG_AIRCON1_HDR_SPACE);
```
```suggestion
      data->iitem(SAMSUNG_AIRCON1_HDR_MARK_US, SAMSUNG_AIRCON1_HDR_SPACE_US);
```
https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L29-L32

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    for (int y = 0; y < 8; y++) {
+      if (send_byte & 0x01) {
+        data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+        data->space(SAMSUNG_AIRCON1_ONE_SPACE);
```
```suggestion
        data->item(SAMSUNG_AIRCON1_BIT_MARK_US, SAMSUNG_AIRCON1_ONE_SPACE_US);
```
https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L29-L32

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+        data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+        data->space(SAMSUNG_AIRCON1_ONE_SPACE);
+      } else {
+        data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+        data->space(SAMSUNG_AIRCON1_ZERO_SPACE);
```
```suggestion
        data->item(SAMSUNG_AIRCON1_BIT_MARK_US, SAMSUNG_AIRCON1_ZERO_SPACE_US);
```
https://github.com/esphome/esphome/blob/387bde665e3ae5d4b8d5a68299abd6032c4291b8/esphome/components/remote_base/remote_base.h#L29-L32

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  data->mark(SAMSUNG_AIRCON1_BIT_MARK);
+  data->space(0);
```
```suggestion
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  auto transmit = this->transmitter_->transmit();
+  auto *data = transmit.get_data();
+  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY);
```
```suggestion
  // Header(2) +  2 * MSG_HDR(2 * Item(2)) + 21 Bytes * 8 Bits * Bit(2) + Last Mark (1)
  data->reserve(2 + 2 * 2 * 2 + 21 * 8 * 2 + 1);
  data->set_carrier_frequency(SAMSUNG_IR_FREQUENCY_HZ);
```

That is 347 items for dynamic size array. Mark or space take up 1 element of space in vector. Item is 2 elements. I think that this is correct but take a look at code and check it. Should be faster at runtime if its reserved. 

> Correctly using reserve() can prevent unnecessary memory reallocations.
https://en.cppreference.com/w/cpp/container/vector/reserve

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  data->mark(SAMSUNG_AIRCON1_BIT_MARK);
```
```suggestion
  data->mark(SAMSUNG_AIRCON1_BIT_MARK_US);
```

jorofi replied:
Is there a specific reason for adding a _US suffix after this variable name?

DjordjeMandic replied:
Only clarity and readability. Easier to maintain code in future. No other practical reason.

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,204 @@
+
+#define GETBITS8(data, offset, size) (((data) & (((uint8_t) UINT8_MAX >> (8 - (size))) << (offset))) >> (offset))
+
+static const char *const TAG = "samsung.climate";
```
```suggestion
```
This should be in source file.

DjordjeMandic reviewed (changes requested):
Using `mark` and `space` separately is fine but if they together mean something like `item` then code is more readable with call to `item`. Is just basic stuff but makes code more readable in my opinion. Also I would suggest reserving enough space. Units in constant names also mean a lot when reading the code. 

One thing that I am not sure myself is about comments, maybe they should be in header instead of source file? At least I would put comments in header. Your comments are not huge, they're medium in size, shouldn't take up too much space in header. Doxygen makes header readable anyways by stripping the comments when looking at source.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,238 @@
+
+namespace esphome {
+namespace samsung {
+
```
```suggestion

static const char *const TAG = "samsung.climate";

```

This is unused, is it needed?

jorofi replied:
I have removed it

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  for (int i = 0; i < 21; i++) {
+    if (i == 7 || i == 14) {
+      data->item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ONE_SPACE);
```
![image](https://github.com/user-attachments/assets/fa1b0b09-341a-41b8-aee4-00d814722913)
Just wanted to check: was using SAMSUNG_AIRCON1_ONE_SPACE intentional, or could it have been an accidental change? If it behaves as expected, then no issues.

jesserockz commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+void SamsungClimate::transmit_state() {
+  if (current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode == climate::ClimateMode::CLIMATE_MODE_OFF) {
+    send_power_state_(false);
```
> All uses of class members and member functions should be prefixed with `this->` to distinguish them from global functions/variables.

https://esphome.io/guides/contributing#codebase-standards
```suggestion
    this->send_power_state_(false);
```

jesserockz commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,228 @@
+    uint8_t : 8;
+    // Byte 5
+    uint8_t : 4;
+    uint8_t Sleep5 : 1;
```
> Function, method and variable names are `lower_snake_case`

https://esphome.io/guides/contributing#codebase-standards
```suggestion
    uint8_t sleep_5 : 1;
```

jorofi commented:
@jesserockz, @DjordjeMandic  I have made the necessary changes. Can you recheck it? Many thanks for considering my request.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+namespace samsung {
+
+void SamsungClimate::transmit_state() {
+  if (current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
```
```suggestion
  if (this->current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+}
+
+void SamsungClimate::set_temp_(const uint8_t temp) {
+  protocol_.temp = esphome::clamp<uint8_t>(temp, K_SAMSUNG_AC_MIN_TEMP, K_SAMSUNG_AC_MAX_TEMP);
```
```suggestion
  this->protocol_.temp = esphome::clamp<uint8_t>(temp, K_SAMSUNG_AC_MIN_TEMP, K_SAMSUNG_AC_MAX_TEMP);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+                                                                    0x01, 0xD2, 0x0F, 0x00, 0x00, 0x00, 0x00,
+                                                                    0x01, 0x02, 0xFF, 0x71, 0x80, 0x11, 0xC0};
+
+  std::memcpy(protocol_.raw, on ? K_ON : K_OFF, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
+
+  this->send_();
+
+  std::memcpy(protocol_.raw, K_RESET, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
```
```suggestion
  std::memcpy(this->protocol_.raw, on ? K_ON : K_OFF, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);

  this->send_();

  std::memcpy(this->protocol_.raw, K_RESET, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+void SamsungClimate::set_temp_(const uint8_t temp) {
+  protocol_.temp = esphome::clamp<uint8_t>(temp, K_SAMSUNG_AC_MIN_TEMP, K_SAMSUNG_AC_MAX_TEMP);
+}
+
+void SamsungClimate::send_power_state_(const bool on) {
+  static const uint8_t K_ON[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH] = {0x02, 0x92, 0x0F, 0x00, 0x00, 0x00, 0xF0,
+                                                                   0x01, 0xD2, 0x0F, 0x00, 0x00, 0x00, 0x00,
+                                                                   0x01, 0xE2, 0xFE, 0x71, 0x80, 0x11, 0xF0};
+
+  static const uint8_t K_OFF[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH] = {0x02, 0xB2, 0x0F, 0x00, 0x00, 0x00, 0xC0,
+                                                                    0x01, 0xD2, 0x0F, 0x00, 0x00, 0x00, 0x00,
+                                                                    0x01, 0x02, 0xFF, 0x71, 0x80, 0x11, 0xC0};
+
+  std::memcpy(protocol_.raw, on ? K_ON : K_OFF, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
+
+  this->send_();
+
+  std::memcpy(protocol_.raw, K_RESET, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
+}
+
+void SamsungClimate::set_fan_(const climate::ClimateFanMode fan_mode) {
+  switch (fan_mode) {
+    case climate::ClimateFanMode::CLIMATE_FAN_LOW:
+      protocol_.fan = K_SAMSUNG_AC_FAN_LOW;
+      break;
+    case climate::ClimateFanMode::CLIMATE_FAN_MEDIUM:
+      protocol_.fan = K_SAMSUNG_AC_FAN_MED;
+      break;
+    case climate::ClimateFanMode::CLIMATE_FAN_HIGH:
+      protocol_.fan = K_SAMSUNG_AC_FAN_HIGH;
+      break;
+    case climate::ClimateFanMode::CLIMATE_FAN_AUTO:
+    default:
+      protocol_.fan = K_SAMSUNG_AC_FAN_AUTO;
+      break;
+  }
```
```suggestion
  switch (fan_mode) {
    case climate::ClimateFanMode::CLIMATE_FAN_LOW:
      this->protocol_.fan = K_SAMSUNG_AC_FAN_LOW;
      break;
    case climate::ClimateFanMode::CLIMATE_FAN_MEDIUM:
      this->protocol_.fan = K_SAMSUNG_AC_FAN_MED;
      break;
    case climate::ClimateFanMode::CLIMATE_FAN_HIGH:
      this->protocol_.fan = K_SAMSUNG_AC_FAN_HIGH;
      break;
    case climate::ClimateFanMode::CLIMATE_FAN_AUTO:
    default:
      this->protocol_.fan = K_SAMSUNG_AC_FAN_AUTO;
      break;
  }
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+namespace samsung {
+
+void SamsungClimate::transmit_state() {
+  if (current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode == climate::ClimateMode::CLIMATE_MODE_OFF) {
+    this->send_power_state_(false);
+    return;
+  }
+
+  if (current_climate_mode_ == climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode != climate::ClimateMode::CLIMATE_MODE_OFF) {
+    this->send_power_state_(true);
+  }
+
+  current_climate_mode_ = this->mode;
```
```suggestion
  if (this->current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
      this->mode == climate::ClimateMode::CLIMATE_MODE_OFF) {
    this->send_power_state_(false);
    return;
  }

  if (this->current_climate_mode_ == climate::ClimateMode::CLIMATE_MODE_OFF &&
      this->mode != climate::ClimateMode::CLIMATE_MODE_OFF) {
    this->send_power_state_(true);
  }

  this->current_climate_mode_ = this->mode;
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+}
+
+void SamsungClimate::set_swing_(const climate::ClimateSwingMode swing_mode) {
+  switch (swing_mode) {
+    case climate::ClimateSwingMode::CLIMATE_SWING_BOTH:
+      protocol_.swing = K_SAMSUNG_AC_SWING_BOTH;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_HORIZONTAL:
+      protocol_.swing = K_SAMSUNG_AC_SWING_H;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_VERTICAL:
+      protocol_.swing = K_SAMSUNG_AC_SWING_V;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_OFF:
+    default:
+      protocol_.swing = K_SAMSUNG_AC_SWING_OFF;
+      break;
+  }
```
```suggestion
  switch (swing_mode) {
    case climate::ClimateSwingMode::CLIMATE_SWING_BOTH:
      this->protocol_.swing = K_SAMSUNG_AC_SWING_BOTH;
      break;
    case climate::ClimateSwingMode::CLIMATE_SWING_HORIZONTAL:
      this->protocol_.swing = K_SAMSUNG_AC_SWING_H;
      break;
    case climate::ClimateSwingMode::CLIMATE_SWING_VERTICAL:
      this->protocol_.swing = K_SAMSUNG_AC_SWING_V;
      break;
    case climate::ClimateSwingMode::CLIMATE_SWING_OFF:
    default:
      this->protocol_.swing = K_SAMSUNG_AC_SWING_OFF;
      break;
  }
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+}
+
+void SamsungClimate::set_swing_(const climate::ClimateSwingMode swing_mode) {
+  switch (swing_mode) {
+    case climate::ClimateSwingMode::CLIMATE_SWING_BOTH:
+      protocol_.swing = K_SAMSUNG_AC_SWING_BOTH;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_HORIZONTAL:
+      protocol_.swing = K_SAMSUNG_AC_SWING_H;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_VERTICAL:
+      protocol_.swing = K_SAMSUNG_AC_SWING_V;
+      break;
+    case climate::ClimateSwingMode::CLIMATE_SWING_OFF:
+    default:
+      protocol_.swing = K_SAMSUNG_AC_SWING_OFF;
+      break;
+  }
+}
+
+void SamsungClimate::set_mode_(const climate::ClimateMode climate_mode) {
+  switch (climate_mode) {
+    case climate::ClimateMode::CLIMATE_MODE_HEAT:
+      protocol_.mode = K_SAMSUNG_AC_HEAT;
+      break;
+    case climate::ClimateMode::CLIMATE_MODE_DRY:
+      protocol_.mode = K_SAMSUNG_AC_DRY;
+      break;
+    case climate::ClimateMode::CLIMATE_MODE_COOL:
+      protocol_.mode = K_SAMSUNG_AC_COOL;
+      break;
+    case climate::ClimateMode::CLIMATE_MODE_FAN_ONLY:
+      protocol_.mode = K_SAMSUNG_AC_FAN;
+      break;
+    case climate::ClimateMode::CLIMATE_MODE_HEAT_COOL:
+    case climate::ClimateMode::CLIMATE_MODE_AUTO:
+    default:
+      protocol_.mode = K_SAMSUNG_AC_AUTO;
+      break;
+  }
```
```suggestion
  switch (climate_mode) {
    case climate::ClimateMode::CLIMATE_MODE_HEAT:
      this->protocol_.mode = K_SAMSUNG_AC_HEAT;
      break;
    case climate::ClimateMode::CLIMATE_MODE_DRY:
      this->protocol_.mode = K_SAMSUNG_AC_DRY;
      break;
    case climate::ClimateMode::CLIMATE_MODE_COOL:
      this->protocol_.mode = K_SAMSUNG_AC_COOL;
      break;
    case climate::ClimateMode::CLIMATE_MODE_FAN_ONLY:
      this->protocol_.mode = K_SAMSUNG_AC_FAN;
      break;
    case climate::ClimateMode::CLIMATE_MODE_HEAT_COOL:
    case climate::ClimateMode::CLIMATE_MODE_AUTO:
    default:
      this->protocol_.mode = K_SAMSUNG_AC_AUTO;
      break;
  }
```

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,228 @@
+class SamsungClimate : public climate_ir::ClimateIR {
+  SamsungProtocol protocol_;
+  climate::ClimateMode current_climate_mode_;
+
+ public:
```
```suggestion
class SamsungClimate : public climate_ir::ClimateIR {
 public:
```

DjordjeMandic replied:
Since no access specifier (e.g., public, protected, or private) is explicitly declared before these members, they will default to private because the default access level in a class is private in C++.

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,228 @@
+                              {climate::CLIMATE_SWING_OFF, climate::CLIMATE_SWING_VERTICAL,
+                               climate::CLIMATE_SWING_HORIZONTAL, climate::CLIMATE_SWING_BOTH}) {}
+
+ protected:
```
```suggestion
 protected:
  SamsungProtocol protocol_;
  climate::ClimateMode current_climate_mode_;

```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,204 @@
+}
+
+void SamsungClimate::checksum_() {
+  uint8_t sectionsum = calc_section_checksum(protocol_.raw);
+  protocol_.sum_1_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
+  protocol_.sum_1_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
+  sectionsum = calc_section_checksum(protocol_.raw + K_SAMSUNG_AC_SECTION_LENGTH);
+  protocol_.sum_2_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
+  protocol_.sum_2_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
+  sectionsum = calc_section_checksum(protocol_.raw + K_SAMSUNG_AC_SECTION_LENGTH * 2);
+  protocol_.sum_3_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
+  protocol_.sum_3_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
```
```suggestion
  uint8_t sectionsum = calc_section_checksum(this->protocol_.raw);
  this->protocol_.sum_1_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
  this->protocol_.sum_1_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
  sectionsum = calc_section_checksum(this->protocol_.raw + K_SAMSUNG_AC_SECTION_LENGTH);
  this->protocol_.sum_2_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
  this->protocol_.sum_2_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
  sectionsum = calc_section_checksum(this->protocol_.raw + K_SAMSUNG_AC_SECTION_LENGTH * 2);
  this->protocol_.sum_3_upper = GETBITS8(sectionsum, K_HIGH_NIBBLE, K_NIBBLE_SIZE);
  this->protocol_.sum_3_lower = GETBITS8(sectionsum, K_LOW_NIBBLE, K_NIBBLE_SIZE);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+  for (uint8_t i = 0; i < 14; i++) {
+    if (i == 7) {
+      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_MSG_SPACE)) {
+        ESP_LOGD(TAG, "Received MSG_SPACE %d", i);
+      } else {
+        ESP_LOGD(TAG, "Failed to receive MSG_SPACE %d", i);
+      }
+
+      if (data.expect_item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE)) {
+        ESP_LOGD(TAG, "Received HDR_SPACE %d", i);
+      } else {
+        ESP_LOGD(TAG, "Failed to receive HDR_SPACE %d", i);
```
```suggestion
        ESP_LOGV(TAG, "Received MSG_SPACE %" PRIu8, i);
      } else {
        ESP_LOGW(TAG, "Failed to receive MSG_SPACE %" PRIu8, i);
        return false;
      }

      if (data.expect_item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE)) {
        ESP_LOGV(TAG, "Received HDR_SPACE %" PRIu8, i);
      } else {
        ESP_LOGW(TAG, "Failed to receive HDR_SPACE %" PRIu8, i);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+      }
+    }
+
+    for (uint8_t y = 0; y < 8; y++) {
+      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ONE_SPACE)) {
+        protocol_.raw[i] |= 1 << y;
+      } else if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ZERO_SPACE)) {
+        protocol_.raw[i] &= ~(1 << y);
+      } else {
+        ESP_LOGD(TAG, "Failed to receive bit %d, %d", i, y);
+      }
```
```suggestion
      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ONE_SPACE)) {
        this->protocol_.raw[i] |= 1 << y;
      } else if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ZERO_SPACE)) {
        this->protocol_.raw[i] &= ~(1 << y);
      } else {
        ESP_LOGW(TAG, "Failed to receive bit %" PRIu8 ", %" PRIu8, i, y);
      }
```

DjordjeMandic replied:
```suggestion
      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ONE_SPACE)) {
        this->protocol_.raw[i] |= 1 << y;
      } else if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ZERO_SPACE)) {
        this->protocol_.raw[i] &= ~(1 << y);
      } else {
        ESP_LOGW(TAG, "Failed to receive bit %" PRIu8 " in byte %" PRIu8, y, i);
        return false;
      }
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+    }
+  }
+
+  ESP_LOGD(TAG, "update_swing_mode_");
+  this->update_swing_mode_();
+
+  ESP_LOGD(TAG, "update_climate_mode_");
+  this->update_climate_mode_();
+
+  ESP_LOGD(TAG, "update_temp_");
+  this->update_temp_();
+
+  ESP_LOGD(TAG, "update_fan_");
+  this->update_fan_();
+
+  ESP_LOGD(TAG, "update_power_");
+  this->update_power_();
+
+  ESP_LOGD(TAG, "publish_state");
+  this->publish_state();
```
```suggestion
  ESP_LOGVV(TAG, "update_swing_mode_");
  this->update_swing_mode_();

  ESP_LOGVV(TAG, "update_climate_mode_");
  this->update_climate_mode_();

  ESP_LOGVV(TAG, "update_temp_");
  this->update_temp_();

  ESP_LOGVV(TAG, "update_fan_");
  this->update_fan_();

  ESP_LOGVV(TAG, "update_power_");
  this->update_power_();

  ESP_LOGD(TAG, "Receive successful, publishing new state.");
  this->publish_state();
```

DjordjeMandic replied:
```suggestion
  this->update_swing_mode_();
  this->update_climate_mode_();
  this->update_temp_();
  this->update_fan_();
  this->update_power_();

  ESP_LOGD(TAG, "Reception successful, publishing new state.");
  this->publish_state();
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+      data->item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE);
+    }
+
+    uint8_t send_byte = protocol_.raw[i];
```
```suggestion
    uint8_t send_byte = this->protocol_.raw[i];
```

DjordjeMandic replied:
```suggestion
    uint8_t send_byte = this->protocol_.raw[i];
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+}
+
+void SamsungClimate::update_power_() {
+  ESP_LOGD(TAG, "power_1: 0x%02X, power_2: 0x%02X", this->protocol_.power_1, this->protocol_.power_2);
```
```suggestion
  ESP_LOGV(TAG, "power_1: 0x%02" PRIX8 ", power_2: 0x%02" PRIX8, this->protocol_.power_1, this->protocol_.power_2);
```

DjordjeMandic replied:
```suggestion
  ESP_LOGV(TAG, "power_1: 0x%02" PRIX8 ", power_2: 0x%02" PRIX8, this->protocol_.power_1, this->protocol_.power_2);
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+  }
+}
+
+void SamsungClimate::update_swing_mode_() {
+  switch (this->protocol_.swing) {
```
```suggestion
void SamsungClimate::update_swing_mode_() {
  ESP_LOGVV(TAG, "update_swing_mode_");
  switch (this->protocol_.swing) {
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+
+  ESP_LOGD(TAG, "Received Samsung A/C message");
+  for (uint8_t i = 0; i < 14; i++) {
+    if (i == 7) {
+      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_MSG_SPACE)) {
+        ESP_LOGD(TAG, "Received MSG_SPACE %d", i);
+      } else {
+        ESP_LOGD(TAG, "Failed to receive MSG_SPACE %d", i);
+      }
+
+      if (data.expect_item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE)) {
+        ESP_LOGD(TAG, "Received HDR_SPACE %d", i);
+      } else {
+        ESP_LOGD(TAG, "Failed to receive HDR_SPACE %d", i);
+      }
+    }
```
```suggestion
    if (i == 7) {
      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_MSG_SPACE)) {
        ESP_LOGV(TAG, "Received MSG_SPACE %" PRIu8, i);
      } else {
        ESP_LOGW(TAG, "Failed to receive MSG_SPACE %" PRIu8, i);
        return false;
      }

      if (data.expect_item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE)) {
        ESP_LOGV(TAG, "Received HDR_SPACE %" PRIu8, i);
      } else {
        ESP_LOGW(TAG, "Failed to receive HDR_SPACE %" PRIu8, i);
        return false;
      }
    }
```

DjordjeMandic commented on esphome/components/samsung/samsung.h:
```diff
@@ -194,6 +195,8 @@ class SamsungClimate : public climate_ir::ClimateIR {
   /// @param[in] on true, the AC is on. false, the AC is off.
   void send_power_state_(bool on);
+  /// Check if the AC is off.
+  bool is_power_off();
```
```suggestion
  bool is_power_off_();
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -219,6 +235,8 @@ void SamsungClimate::send_power_state_(const bool on) {
   std::memcpy(this->protocol_.raw, K_RESET, K_SAMSUNG_AC_EXTENDED_STATE_LENGTH);
 }
 
+bool SamsungClimate::is_power_off() { return this->protocol_.power_1 == 0; }
```
```suggestion
bool SamsungClimate::is_power_off_() { return this->protocol_.power_1 == 0; }
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -55,27 +59,37 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
 
-  ESP_LOGD(TAG, "update_climate_mode_");
-  this->update_climate_mode_();
+    ESP_LOGD(TAG, "Reception successful, power is off, publishing new state.");
+    this->publish_state();
+    return true;
```
```suggestion
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -55,27 +59,37 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
 
-  ESP_LOGD(TAG, "update_fan_");
-  this->update_fan_();
+    ESP_LOGD(TAG, "Reception successful, power is on, publishing new state.");
+    this->publish_state();
+    return true;
```
```suggestion
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -55,27 +59,37 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
 
-  ESP_LOGD(TAG, "update_temp_");
-  this->update_temp_();
+  if (this->last_known_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF) {
```
```suggestion
  bool last_known_mode_is_not_off = this->last_known_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF;
  if (last_known_mode_is_not_off) {
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -55,27 +59,37 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
+  this->update_fan_();
 
-  ESP_LOGD(TAG, "publish_state");
+  ESP_LOGD(TAG, "Reception successful, publishing new state.");
```
```suggestion
  ESP_LOGD(TAG, "Reception successful, power is %s, publishing new state.", ONOFF(!this->is_power_off_()));
```

DjordjeMandic replied:
I'd avoid multiple log messages that are the same but in different place. Have single call to {publish state and return true} if you need to log before it. Use if/else to prepare before publishing state.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -55,27 +59,37 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
 
-  ESP_LOGD(TAG, "update_power_");
-  this->update_power_();
+  this->update_climate_mode_();
+  this->update_swing_mode_();
+  this->update_temp_();
+  this->update_fan_();
```
```suggestion
  if (!(this->is_power_off_() && last_known_mode_is_not_off)) {
    this->update_climate_mode_();
    this->update_swing_mode_();
    this->update_temp_();
    this->update_fan_();
  }
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -1,4 +1,5 @@
 #include "samsung.h"
+#include <cinttypes>
```
I'm using same definitions from `inttypes.h` in HX711 yet it compiles. I think that this is not needed.
```suggestion
```

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+      if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_MSG_SPACE)) {
+        ESP_LOGV(TAG, "Received MSG_SPACE %" PRIu8, i);
+      } else {
+        ESP_LOGW(TAG, "Failed to receive MSG_SPACE %" PRIu8, i);
```
```suggestion
        ESP_LOGW(TAG, "Failed to receive MSG_SPACE %" PRIu8, i);
        return false;
```
Discard the entire packet due to an error during reception.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+      if (data.expect_item(SAMSUNG_AIRCON1_HDR_MARK, SAMSUNG_AIRCON1_HDR_SPACE)) {
+        ESP_LOGV(TAG, "Received HDR_SPACE %" PRIu8, i);
+      } else {
+        ESP_LOGW(TAG, "Failed to receive HDR_SPACE %" PRIu8, i);
```
```suggestion
        ESP_LOGW(TAG, "Failed to receive HDR_SPACE %" PRIu8, i);
        return false;
```
Discard the entire packet due to an error during reception.

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,325 @@
+      } else if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ZERO_SPACE)) {
+        protocol_.raw[i] &= ~(1 << y);
+      } else {
+        ESP_LOGW(TAG, "Failed to receive bit %" PRIu8 " in byte %" PRIu8, y, i);
```
```suggestion
        ESP_LOGW(TAG, "Failed to receive bit %" PRIu8 " in byte %" PRIu8, y, i);
        return false;
```
Discard the entire packet due to an error during reception.

DjordjeMandic reviewed (approved):
Looks good to me

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,332 @@
+  this->send_();
+
+  if (on) {  // Give time for the AC receiver to process the ON command.
+    delay(1000);
```
https://esphome.io/guides/contributing.html#a-note-about-delays-in-code

Try to find workaround. State machine or something else. Could you explain the problem?

jorofi replied:
I will change on_receive to handle the message better and will remove this delay.

jorofi replied:
@DjordjeMandic I have found the root cause and I have removed this delay completely.

DjordjeMandic reviewed (changes requested):
No delays above 10ms allowed

DjordjeMandic commented:
Try managing unimportant delays with https://esphome.io/api/classesphome_1_1_component#af8a76dc985ac472662d1c2165624f72f

DjordjeMandic commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -37,7 +32,7 @@ bool SamsungClimate::on_receive(remote_base::RemoteReceiveData data) {
   }
 
-  ESP_LOGD(TAG, "Received Samsung A/C message");
+  ESP_LOGD(TAG, "Received Samsung A/C message size %" PRIu8, data.size());
```
```suggestion
  ESP_LOGD(TAG, "Received Samsung A/C message size %" PRId32, data.size());
```
https://github.com/esphome/esphome/blob/dbf4c2c4dae43ee929471a26823db88f373576d6/esphome/components/remote_base/remote_base.h#L56

DjordjeMandic reviewed (changes requested):
https://en.cppreference.com/w/c/types/integer#Format_macro_constants

Using "%" PRId32 for formatting int32_t variables ensures compatibility across different platforms and compilers. While it might look less clean than "%d", it avoids potential issues caused by platform-specific variations in type sizes. It's considered good practice to use format specifier macros like PRId32 whenever possible, as they provide a standardized and portable way to handle integer types. For more details, see the related discussion in PR #8122

jorofi commented:
Hi, @jesserockz can you merge this PR? I have addressed all the open issues.

DjordjeMandic commented:
> Hi, @jesserockz can you merge this PR? I have addressed all the open issues.

https://esphome.io/guides/contributing.html#review-by-maintainers
Someone will review it. 

One thing that I noticed was that you had open PR while still pushing new changes. Maybe devs understood that its still WIP and didn't review it. 
https://esphome.io/guides/contributing.html#submitting-a-pull-request

esnet146 commented:
Tested on air conditioner model AQ07XAN and chip esp8266.
The result is excellent!

ximex reviewed (changes requested):
update the test files like in this PR (https://github.com/esphome/esphome/pull/9178/files) with the include syntax

DjordjeMandic reviewed (approved):
LGTM

swoboda1337 commented on tests/components/samsung/common.yaml:
```diff
@@ -0,0 +1,7 @@
+
+climate:
+  - platform: samsung
+    name: SAMSUNG Climate
```
Tests need to be updated after recent changes. Thanks

jorofi replied:
Hi @swoboda1337, the tests are updated

jorofi committed:
Add support for Samsung climate IR

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

swoboda1337 committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
Fix invalid frame tail by sending correct section count for normal vs extended messages

jorofi commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,241 @@
+const uint8_t K_HIGH_NIBBLE = 4;
+
+static const uint8_t K_RESET[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH] = {0x02, 0x92, 0x0F, 0x00, 0x00, 0x00, 0xF0,
+                                                                    0x01, 0x02, 0xAE, 0x71, 0x00, 0x15, 0xF0};
```
This issue is resolved

jorofi committed:
Fix temperature decoding and invalid frame tail on transmit

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jorofi commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,324 @@
+      esphome::clamp<uint8_t>(temp, K_SAMSUNG_AC_MIN_TEMP, K_SAMSUNG_AC_MAX_TEMP) - K_SAMSUNG_AC_MIN_TEMP;
+}
+
+void SamsungClimate::update_temp_() { this->target_temperature = this->protocol_.temp; }
```
The issue is resolved, and the temperature is correctly set. Tested on Samsung AC AR12RXFPEWQNEU

jorofi committed:
Uninitialized member fields current_climate_mode_, and last_known_mode_

jorofi committed:
Merge branch 'component_climate_ir_samsung' of https://github.com/jorofi/esphome into component_climate_ir_samsung

jorofi commented on esphome/components/samsung/samsung.h:
```diff
@@ -0,0 +1,241 @@
+ protected:
+  SamsungProtocol protocol_;
+  climate::ClimateMode current_climate_mode_;
+  climate::ClimateMode last_known_mode_;
```
The change is applied and tested on Samsung AC AR12RXFPEWQNEU

jorofi committed:
Fully initializes all 21 bytes with the correct Samsung AC reset state

jorofi committed:
current_climate_mode_ is set to OFF before sending the power-off frame, so subsequent transmit_state() calls won't re-enter the OFF-transition branch and internal state stays consistent.

jorofi commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,324 @@
+
+void SamsungClimate::transmit_state() {
+  if (this->current_climate_mode_ != climate::ClimateMode::CLIMATE_MODE_OFF &&
+      this->mode == climate::ClimateMode::CLIMATE_MODE_OFF) {
+    this->last_known_mode_ = this->current_climate_mode_;
+    this->send_power_state_(false);
```
Tested and resolved

jorofi committed:
The receive path now reads the full extended frame

jorofi commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,324 @@
+
+  ESP_LOGD(TAG, "Received Samsung A/C message size %" PRId32, data.size());
+  for (uint8_t i = 0; i < 14; i++) {
+    if (i == 7) {
```
Tested and fixed

jorofi committed:
Add remote_receiver to tests

jorofi committed:
Fixed styling

jorofi committed:
validate the trailing footer mark

jorofi commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,324 @@
+        protocol_.raw[i] |= 1 << y;
+      } else if (data.expect_item(SAMSUNG_AIRCON1_BIT_MARK, SAMSUNG_AIRCON1_ZERO_SPACE)) {
+        protocol_.raw[i] &= ~(1 << y);
+      } else {
+        ESP_LOGW(TAG, "Failed to receive bit %" PRIu8 " in byte %" PRIu8, y, i);
+        return false;
+      }
+    }
+  }
+
+  if (this->is_power_off_()) {
```
The validation is added and tested

jorofi committed:
resolving the bugprone-too-small-loop-variable

jorofi commented on esphome/components/samsung/samsung.cpp:
```diff
@@ -0,0 +1,324 @@
+  this->current_climate_mode_ = this->mode;
+
+  this->set_climate_mode_(this->mode);
+  this->set_temp_(this->target_temperature);
```
Done

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
use this-> on all places. Use inline get_bits8

jorofi committed:
Fixed. The K_RESET array in samsung.h:53-55 now restores the working section 2 (bytes 7‑13) while staying fully 21‑byte initialized.

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung

jorofi committed:
Merge branch 'dev' into component_climate_ir_samsung
